pkg/utils/environment/kernel_symbols.go

```diff
@@ -2,303 +2,223 @@ package environment
+// GetSymbolByOwnerAndName returns all the symbols with the given owner and name.
+func (kst *KernelSymbolTable) GetSymbolByOwnerAndName(owner, name string) ([]*KernelSymbol, error) {
+	symbols, err := kst.symbols.LookupByName(name)
 	if err != nil {
-		return err
+		return nil, errfmt.WrapError(err)
 	}
-	defer func() {
-		_ = file.Close()
-	}()
 
-	// Read the kallsyms file line by line and process each line.
-	scanner := bufio.NewScanner(file)
-	for scanner.Scan() {
-		fields := strings.Fields(scanner.Text())
-		if len(fields) < 3 {
-			continue
-		}
-		sym := parseKallsymsLine(fields)
-		if sym == nil {
-			continue
+	// Find symbols that have the requested owner
+	filteredSymbols := []*KernelSymbol{}
+	for _, symbol := range symbols {
+		if symbolOwners[symbol.owner] == owner {
+			filteredSymbols = append(filteredSymbols, symbol)
 		}
-
-		if k.onlyRequired {
-			_, symRequired := k.requiredSyms[sym.Name]
-			_, addrRequired := k.requiredAddrs[sym.Address]
-			if !symRequired && !addrRequired {
-				continue
-			}
-		}
-
-		k.symbols[sym.Name] = append(k.symbols[sym.Name], sym)
-		k.addrs[sym.Address] = append(k.addrs[sym.Address], sym)
-		k.symByName[nameAndOwner{sym.Name, sym.Owner}] = append(k.symByName[nameAndOwner{sym.Name, sym.Owner}], sym)
-		k.symByAddr[addrAndOwner{sym.Address, sym.Owner}] = append(k.symByAddr[addrAndOwner{sym.Address, sym.Owner}], sym)
 	}
-	err = scanner.Err()
 
-	return err
+	return filteredSymbols, nil
 }
 
-// parseKallsymsLine parses a line from /proc/kallsyms and returns a KernelSymbol.
-func parseKallsymsLine(line []string) *KernelSymbol {
-	if len(line) < 3 {
-		return nil
-	}
+// GetSymbolByAddr returns all the symbols with the given address.
+func (kst *KernelSymbolTable) GetSymbolByAddr(addr uint64) ([]*KernelSymbol, error) {
+	symbols, err := kst.symbols.LookupByAddressExact(addr)
+	return symbols, errfmt.WrapError(err)
+}
 
-	symbolAddr, err := strconv.ParseUint(line[0], 16, 64)
+// GetPotentiallyHiddenSymbolByAddr returns all the symbols with the given address,
+// or if none are found, a fake symbol with the "hidden" owner.
+func (kst *KernelSymbolTable) GetPotentiallyHiddenSymbolByAddr(addr uint64) []*KernelSymbol {
+	symbols, err := kst.symbols.LookupByAddressExact(addr)
 	if err != nil {
-		return nil
+		return []*KernelSymbol{{
+			address: addr,
+			owner:   symbolOwnerToIdx["hidden"],
+		}}
 	}
 
-	symbolType := line[1]
-	symbolName := line[2]
-
-	symbolOwner := "system"
-	if len(line) > 3 {
-		line[3] = strings.TrimPrefix(line[3], "[")
-		line[3] = strings.TrimSuffix(line[3], "]")
-		symbolOwner = line[3]
-	}
+	return symbols
+}
 
-	return &KernelSymbol{
-		Name:    symbolName,
-		Type:    symbolType,
-		Address: symbolAddr,
-		Owner:   symbolOwner,
-	}
+func (kst *KernelSymbolTable) ForEachSymbol(callback func(*KernelSymbol)) {
+	kst.symbols.ForEachSymbol(callback)
 }
 
 // copySliceOfPointersToSliceOfStructs converts a slice of pointers to a slice of structs.
```

Comment:
As this is left unused inside the package, and I see you've copied it elsewhere, this can be removed, no reason to leave dangling code.

Reply:
It's used in the test, should I remove it from the package and call the inner version in the test?

Reply:
I think you've copied it into the test file no? Then you're using that copy anyway and this is dangling.

Reply:
No, `getTheOnlySymbol()` in the test file uses `KernelSymbolTable.ForEachSymbol()`

Reply:
I'm referring to `copySliceOfPointersToSliceOfStructs `, I see now it might not be clear from the snippet.

Reply:
Oh you're right, I missed that.